src/tools/miri/tests/pass-dep/shims/posix_memalign.rs

```diff
@@ -37,7 +37,7 @@ fn main() {
         let size = 31;
         assert_eq!(libc::posix_memalign(&mut ptr, align, size), 0);
         assert!(!ptr.is_null());
-        assert!(ptr.is_aligned_to(align));
```

Comment:
This does seem like the function can be quite nice to have, doesn't it?

And if it's used heavily inside the standard library (and hence can't be removed), that is also a data point in favor of it actually being useful. So I don't quite see why it should get removed.

Reply:
It seems to only be used for asserts. Maybe we should have a new assert macro?

Reply:
In general, an application developer probably wants `.is_aligned_to::<N>`. It is rare that you truly want a dynamic alignment check.

Reply:
so it's only used heavily in the stdlib at this point because:

* people not being aware how powerful `ptr.cast::<T>().some_check()` is as an idiom
* assert_unsafe_precondition was added in a ton of places calling is_aligned_and_not_null which calls this

The former I rewrote, the latter is a bit ??? to me. The functions/macros are clearly written in a very specific style that avoids generics, and I can't tell if this is Important For Something or someone being idiosyncratic for no good reason.

Reply:
Sorry just to clarify: as used is_aligned_and_not_null could just be changed to be generic and call .is_aligned(). But I'm not sure if that would break some magic.

Reply:
@saethlin  carefully made it all non-generic to reduce codegen effort and avoid the build performance regressions we'd otherwise see.

Reply:
That makes total sense, although it's a bit idiosyncratic as a public API.

Reply:
also Extremely funny saethlin is the one who suggested removing it X3

Reply:
:raised_eyebrow: I don't and didn't mean that we should *never* stabilize such a function. But at the current time, I think we have evidence that this meets the bar for what goes in the standard library.

In particular, I find the API of this function quite vexing. The `usize` argument raises annoying questions because not every `usize` is a valid alignment. The design needs to commit to one of

* Don't try to detect alignments that aren't a power of 2. I dislike this option because we shouldn't let trivial-to-detect bugs go unnoticed. A debug assertion would get compiled out of the distributed standard library and therefore not really exist for most users so I'm not really in favor of that.
* Panic when passed an alignment that's not a power of 2. I dislike this option because generates a substantial amount of IR and I expect that if this function is used by performance nerds like me the minor inefficiency will invite people to roll their own and advocate against using the standard library function. At which point, why are we stabilizing something people will replace in a third-party library.
* Make the alignment a `const` param like Jubilee says above. This would rule out 2 of the 3 users of a function called `is_aligned_to` that I linked before. So I'm not particularly convinced that there are many users of a `const ALIGN` version, considering that the 1 user who could use such an API can also use `is_aligned::<T>()`.
* Change the argument type to an `Alignment`. This seems to me like the right way to design a Rust API, because it rules out invalid usage with the type system. This option also increases the difficulty of stabilizing something, but that seems fine when the other options are stabilizing something with a high chance of regret.

Reply:
To be clear I think it's perfectly fine to stabilize `is_aligned` (which checks the pointee type, right?) and leave `is_aligned_to` unstable, if there is consensus that `is_aligned` is sufficiently useful and will remain useful even if `is_aligned_to` eventually becomes stable. That would require moving them to different feature gates.

I am just surprised about the proposal of making them private. That's a pretty rare thing to be done with a unstable std API I think. Usually it either gets removed or stabilized.

Reply:
Ah I guess you're already doing that in https://github.com/rust-lang/rust/pull/121948. :)

Reply:
I'll add my 2¢: I think `is_aligned_to` is useful (even if much less than plain `is_aligned`), as suggested by the uses in std and other potential uses in the ecosystem. But I also agree that accepting `usize` is not nice.

IMO we should change it to accepting `Alignment` and stabilize `is_aligned` separately.